packages/gatsby-plugin-sharp/src/scheduler.js

```diff
@@ -1,88 +1,140 @@
 const _ = require(`lodash`)
 const { existsSync } = require(`fs`)
 const uuidv4 = require(`uuid/v4`)
-const queue = require(`async/queue`)
-const { processFile } = require(`./process-file`)
+const pDefer = require(`p-defer`)
+const worker = require(`./worker`)
 const { createProgress } = require(`./utils`)
 
-const toProcess = {}
-let totalJobs = 0
-const q = queue((task, callback) => {
-  task(callback)
-}, 1)
+const toProcess = new Map()
+let imagesToProcess = 0
+let imagesFinished = 0
 
 let bar
-// when the queue is empty we stop the progressbar
-q.drain = () => {
-  if (bar) {
-    bar.done()
-  }
-  totalJobs = 0
-}
-
-const getFileKey = filePath => filePath.replace(/\./g, `%2E`)
 
-const setJobToProcess = (toProcess, job, deferred) => {
-  const inputFileKey = getFileKey(job.inputPath)
-  const outputFileKey = getFileKey(job.outputPath)
-  const jobPath = `["${inputFileKey}"].["${outputFileKey}"]`
-
-  // Check if the job has already been queued. If it has, there's nothing
-  // to do, return.
-  if (_.has(toProcess, jobPath)) {
-    return { existingPromise: _.get(toProcess, `${jobPath}.deferred.promise`) }
+// node 8 doesn't support promise.finally, we extract this function to re-use it inside then & catch
+const cleanupJob = (job, boundActionCreators) => {
+  if (bar) {
+    bar.tick(job.task.args.operations.length)
   }
 
-  // Check if the output file already exists so we don't redo work.
-  if (existsSync(job.outputPath)) {
-    return { existingPromise: Promise.resolve(job) }
-  }
+  imagesFinished += job.task.args.operations.length
 
-  let isQueued = false
-  if (toProcess[inputFileKey]) {
-    isQueued = true
+  if (imagesFinished === imagesToProcess) {
+    if (bar) {
+      bar.done()
+      bar = null
+    }
+    imagesToProcess = 0
+    imagesFinished = 0
   }
 
-  _.set(toProcess, jobPath, {
-    job: job,
-    deferred,
-  })
-
-  return { isQueued }
+  boundActionCreators.endJob({ id: job.id }, { name: `gatsby-plugin-sharp` })
 }
 
+const executeJobs = _.throttle(
+  (pluginOptions, boundActionCreators) => {
+    toProcess.forEach(job => {
+      const { task } = job
+      toProcess.delete(task.inputPaths[0])
```

Comment:
Let's leave a comment for failure here?

```suggestion
      // We don't add this back even if it fails at the moment 
      toProcess.delete(task.inputPaths[0])
```